src/mame/drivers/centiped.cpp

```diff
@@ -822,36 +823,58 @@ void centiped_state::milliped_map(address_map &map)
  - 74LS245
  - 2*8KB gfx roms, on a separate smaller daughterboard
-
- TODO: centiped does not work yet, the game reconfigures the memorymap
 */
 
 void centiped_state::multiped_map(address_map &map)
 {
+	map(0x0000, 0x7fff).mirror(0x8000).bankr("rombank");
 	map(0x0000, 0x03ff).ram();
-	map(0x0400, 0x040f).rw("pokey", FUNC(pokey_device::read), FUNC(pokey_device::write));
+	map(0x0400, 0x07ff).rw(FUNC(centiped_state::multiped_0xxx_r), FUNC(centiped_state::multiped_0xxx_w));
 	map(0x0800, 0x080f).rw("pokey2", FUNC(pokey_device::read), FUNC(pokey_device::write));
 	map(0x1000, 0x13bf).ram().w(FUNC(centiped_state::centiped_videoram_w)).share("videoram");
 	map(0x13c0, 0x13ff).ram().share("spriteram");
-	map(0x2000, 0x2000).r(FUNC(centiped_state::centiped_IN0_r));
-	map(0x2001, 0x2001).r(FUNC(centiped_state::milliped_IN1_r));
-	map(0x2010, 0x2010).r(FUNC(centiped_state::milliped_IN2_r));
-	map(0x2011, 0x2011).portr("IN3");
-	map(0x2030, 0x2030).nopr();
-	map(0x2480, 0x249f).w(FUNC(centiped_state::milliped_paletteram_w)).share("paletteram");
-	map(0x2500, 0x2507).w("outlatch", FUNC(ls259_device::write_d7));
-	map(0x2600, 0x2600).w(FUNC(centiped_state::irq_ack_w));
-	map(0x2680, 0x2680).w("watchdog", FUNC(watchdog_timer_device::reset_w));
-	map(0x2700, 0x2700).nopw();
-	map(0x2780, 0x27bf).nopw();
-	map(0x4000, 0x5fff).rom();
-	map(0x6000, 0x7fff).mirror(0x8000).rom();
-	map(0x8000, 0xbfff).rom();
+	map(0x1400, 0x140f).rw("pokey", FUNC(pokey_device::read), FUNC(pokey_device::write));
+	map(0x2000, 0x3fff).r(FUNC(centiped_state::multiped_2xxx_r));
+	map(0x2480, 0x249f).mirror(0x8000).w(FUNC(centiped_state::milliped_paletteram_w)).share("paletteram");
+	map(0x2500, 0x2507).mirror(0x8000).w("outlatch", FUNC(ls259_device::write_d7));
+	map(0x2600, 0x2600).mirror(0x8000).w(FUNC(centiped_state::irq_ack_w));
+	map(0x2680, 0x2680).mirror(0x8000).w("watchdog", FUNC(watchdog_timer_device::reset_w));
+	map(0x2700, 0x2700).mirror(0x8000).nopw();
+	map(0x2780, 0x27bf).mirror(0x8000).nopw();
+	map(0xa000, 0xa000).r(FUNC(centiped_state::centiped_IN0_r));
+	map(0xa001, 0xa001).r(FUNC(centiped_state::milliped_IN1_r));
+	map(0xa010, 0xa010).r(FUNC(centiped_state::milliped_IN2_r));
+	map(0xa011, 0xa011).portr("IN3");
+	map(0xa030, 0xa030).nopr();
 	map(0xd000, 0xd7ff).w(FUNC(centiped_state::multiped_eeprom_w));
 	map(0xd800, 0xd800).mirror(0x03ff).rw(FUNC(centiped_state::multiped_eeprom_r), FUNC(centiped_state::multiped_prgbank_w));
 	map(0xdc00, 0xdc00).mirror(0x03ff).w(FUNC(centiped_state::multiped_gfxbank_w));
 }
 
+uint8_t centiped_state::multiped_0xxx_r(offs_t offset)
+{
+	if (!m_prg_bank)
+		return m_maincpu->space(AS_PROGRAM).read_byte(0x1400 | offset);
+	else
+		return m_maincpu->space(AS_PROGRAM).read_byte(0x1000 | offset);
+}
+
+void centiped_state::multiped_0xxx_w(offs_t offset, uint8_t data)
+{
+	if (!m_prg_bank)
+		m_maincpu->space(AS_PROGRAM).write_byte(0x1400 | offset, data);
+	else
+		m_maincpu->space(AS_PROGRAM).write_byte(0x1000 | offset, data);
+}
```

Comment:
You’re better off making `m_prg_bank` the actual offset value (0x1400 or 0x1000) so you can avoid the test here and get better pipeline efficiency.

Reply:
While I'm at it, does this seem like a reasonable alternative to checking the bank number in `multiped_2xxx_r`?

```c
if (!bank)
{
	m_0xxx_base = 0x1400;
	m_maincpu->space(AS_PROGRAM).install_read_handler(0x2000, 0x3fff, read8sm_delegate(*this, FUNC(centiped_state::multiped_2xxx_r)));
}
else
{
	m_0xxx_base = 0x1000;
	m_maincpu->space(AS_PROGRAM).install_rom(0x2000, 0x3fff, (uint8_t*)m_rombank->base() + 0x2000);
}
```

Reply:
It depends on how frequently this happens.  Installing handlers is relatively expensive, so you don’t want to be doing it dozens or hundreds of times per second.  If the bank is changed relatively infrequently (say, a couple of times per second at most), it’s better to do by installing handlers like this.  If it’s changed very frequently (e.g. multiple times every frame interval) it’s better to keep the trampoline.

Reply:
The memory map only changes once each time the player switches games, not while the machine is running normally, so it should be okay. It at least seems better than checking the currently selected bank every time $2000-3fff is accessed, since that's where the original program ROM for Centipede is mapped.